### PR TITLE
FacT-Finetuning SAM

### PR DESCRIPTION
**Pull request by caroteu**
Implementation of FacT Finetuning method on SAM. @anwai98

### CODE REVIEW

File: finetuning/evaluation/evaluate_instance_segmentation.py [hunk 1]
```diff
@@ -35,9 +35,9 @@ def eval_instance_segmentation_with_decoder(dataset_name, prediction_folder, exp
 
 def main():
     args = get_default_arguments()
-
+    peft_kwargs = {"rank": args.peft_rank, "module": args.peft_module}
```
Comment (caroteu) [inline finetuning/evaluation/evaluate_instance_segmentation.py:38]: @anwai98 I changed the way the way the peft arguments are passed in the evaluation because I thought this would make things more flexible for different methods. Not sure if this is the most elegant way to do it

File: finetuning/evaluation/util.py [hunk 2]
```diff
@@ -80,12 +80,10 @@ def get_dataset_paths(dataset_name, split_choice):
     return raw_dir, labels_dir
 
 
-def get_model(model_type, ckpt, lora_rank):
```
Comment (caroteu) [inline finetuning/evaluation/util.py:83]: @anwai98 This function should be unnecessary and can be replaced by directly calling get_sam_model

File: micro_sam/models/peft_sam.py [hunk 3]
```diff
@@ -56,6 +56,49 @@ def forward(self, x):
...
+    ):
+        super().__init__()
+        self.qkv_proj = block.attn.qkv
+        self.dim = self.qkv_proj.in_features 
+
+        self.q_FacTs = nn.Linear(rank, rank, bias=False)
+        self.v_FacTs = nn.Linear(rank, rank, bias=False)
+
+        # NOTE : Dropout is not included in the original implementation
+        self.dp_q = nn.Dropout(0.1)
```
Comment (caroteu) [inline micro_sam/models/peft_sam.py:81]: @anwai98 Let's discuss whether or not it makes sense to keep the dropout layers, they were included in the MA-SAM implementation but not mentioned in the FacT paper

File: micro_sam/models/peft_sam.py [hunk 4]
```diff
@@ -79,6 +122,10 @@ def __init__(
 
         assert rank > 0
 
+        dim = model.image_encoder.blocks[0].attn.qkv.in_features
+        self.FacTu = nn.Linear(dim, rank, bias=False)
```
Comment (caroteu) [inline micro_sam/models/peft_sam.py:126]: @anwai98 Do we need to add a condition here, so that these parameters are only initialized when FacT is used?

**Comment by anwai98** [comment]
I'll close this PR in favor of moving it to https://github.com/computational-cell-analytics/micro-sam/pull/682.

File: finetuning/evaluation/evaluate_instance_segmentation.py (hunk 1 above)
Comment (anwai98) [reply finetuning/evaluation/evaluate_instance_segmentation.py:38]: I agree. This would be the most elegant way to handle the current functionalities.

File: finetuning/evaluation/util.py (hunk 2 above)
Comment (anwai98) [reply finetuning/evaluation/util.py:83]: Yep, makes sense to get rid of it. I took care of it. Will push the changes in the new PR soon.

File: micro_sam/models/peft_sam.py (hunk 4 above)
Comment (anwai98) [reply micro_sam/models/peft_sam.py:126]: Hmm, we definitely cannot hard-code them in the current location, would create a bit of a mess. While reviewing, Ill try to see how we can make it flexible upon the peft module.
